lib/AST/DiagnosticConsumer.cpp

```diff
@@ -165,29 +165,52 @@ FileSpecificDiagnosticConsumer::subconsumerForLocation(SourceManager &SM,
 void FileSpecificDiagnosticConsumer::handleDiagnostic(
     SourceManager &SM, SourceLoc Loc, DiagnosticKind Kind,
     StringRef FormatString, ArrayRef<DiagnosticArgument> FormatArgs,
-    const DiagnosticInfo &Info) {
+    const DiagnosticInfo &Info, StringRef currentPrimaryInput) {
 
   HasAnErrorBeenConsumed |= Kind == DiagnosticKind::Error;
 
-  Optional<FileSpecificDiagnosticConsumer::Subconsumer *> subconsumer;
+  auto subconsumer = findSubconsumerAndRememberItForNotes(SM, Loc, Kind);
+  if (subconsumer.hasValue() && !(*subconsumer)->getConsumer())
+    subconsumer = findSubconsumerForPrimaryCausingErrorInNonprimary(
```

Comment:
I think you need to remember-for-notes in this case too, so I suggest just pushing this into your new `findSubconsumerAndRememberItForNotes`, or even `subconsumerForLocation`.

Reply:
I agree, and I was under the impression that the code in this PR does that. Specifically,`findSubconsumerForPrimaryCausingErrorInNonprimary` returns through `findSubconsumerAndRememberItForNotes`. I see why it was misleading, though I don't have a better alternative.

Reply:
If you think it's clearer, `findSubconsumerForPrimaryCausingErrorInNonprimary` could return a possibly-invalid `SourceLoc`. And the call to `findSubconsumerAndRememberItForNotes` could then be moved up into `handleDiagnostic`. 

Hmm... I have an idea for a better refactoring. Will try it out.

Reply:
Oh, whoops. I read the code and didn't take that in at all because of the name.

Reply:
But your reaction was good because it led to clearer code!